src/main.rs

```diff
@@ -285,77 +286,108 @@ fn main() {
-        };
+            let signal = match next {
+                future::Either::A((signal, _)) => signal.expect("Signal streams never end"),
+                future::Either::B((_crash, _)) => SIGINT, // Trigger graceful shutdown if a component crashed
+            };
 
-        if signal != SIGHUP {
-            break signal;
-        }
+            if signal != SIGHUP {
+                break signal;
+            }
 
-        // Reload config
-        info!(
-            message = "Reloading config.",
-            path = ?opts.config_path
-        );
+            // Reload config
+            info!(
+                message = "Reloading config.",
+                path = ?opts.config_path
+            );
 
-        let file = if let Some(file) = open_config(&opts.config_path) {
-            file
-        } else {
-            continue;
+            let file = if let Some(file) = open_config(&opts.config_path) {
+                file
+            } else {
+                continue;
+            };
+
+            trace!("Parsing config");
+            let config = vector::topology::Config::load(file);
+            let config = handle_config_errors(config);
+            if let Some(config) = config {
+                let success =
+                    topology.reload_config_and_respawn(config, &mut rt, opts.require_healthy);
+                if !success {
+                    error!("Reload was not successful.");
+                }
+            } else {
+                error!("Reload aborted.");
+            }
         };
 
-        trace!("Parsing config");
-        let config = vector::topology::Config::load(file);
-        let config = handle_config_errors(config);
-        if let Some(config) = config {
-            let success = topology.reload_config_and_respawn(config, &mut rt, opts.require_healthy);
-            if !success {
-                error!("Reload was not successful.");
+        if signal == SIGINT || signal == SIGTERM {
+            use futures::future::Either;
+
+            info!("Shutting down.");
+            let shutdown = topology.stop();
+            metrics_trigger.cancel();
+
+            match rt.block_on(shutdown.select2(signals.into_future())) {
+                Ok(Either::A(_)) => { /* Graceful shutdown finished */ }
+                Ok(Either::B(_)) => {
+                    info!("Shutting down immediately.");
+                    // Dropping the shutdown future will immediately shut the server down
+                }
+                Err(_) => unreachable!(),
             }
+        } else if signal == SIGQUIT {
+            info!("Shutting down immediately");
+            drop(topology);
         } else {
-            error!("Reload aborted.");
+            unreachable!();
         }
-    };
+    }
+    #[cfg(windows)]
+    {
+        let mut ctrl_c_stream = tokio_signal::ctrl_c().flatten_stream();
+        let ctrl_c = future::poll_fn(move || ctrl_c_stream.poll());
```

Comment:
Since you want to reuse this stream below, you could do something like

```rust
let (ctrl_c, ctrl_c_stream) = tokio_signal::ctrl_c().flatten_stream().into_future();
```

Reply:
Thanks! I'm going to use it.